### ISSUE TITLE
Automatically understand types in `Seq`/`List` and enums that are children?

### ISSUE THREAD

**Issue by jnatten** [scala-tsi/scala-tsi#176]
This might just be me not understanding how the plugin works, but I was wondering if there was a way to automatically understand the types of my classes that are in `Seq`/`List`'s (Also enums)?

For example if I want to generate the typescript types for the following types (And hopefully only specify `ParentObj` in build.sbt):
```scala
object CoolEnum extends Enumeration {
  val Yep = Value
  val No = Value
}

case class ListObj(a: String, b: String)

case class ParentObj(
  coolList: Seq[ListObj],
  coolEnum: CoolEnum.Value
)
```

It seems that I have to do something like this to make scala-tsi understand the types:
```scala
import com.scalatsi._
import com.scalatsi.dsl._

object TSTypes {
  implicit val listObj: TSType[ListObj] = TSType.fromCaseClass[ListObj]
  implicit val coolEnum: TSType[CoolEnum.type] = TSType.scalaEnumTSType[CoolEnum.type]
  implicit val parentObj: TSIType[ParentObj] = {
    implicit val xx: TSType[CoolEnum.Value] = TSType.external("CoolEnum")
    TSType.fromCaseClass[ParentObj]
  }
}
```

If i remove the line about `listObj` (Which is the most common usecase we're running into. I'll live pretty good with handling the enums manually :smile: ) I get the following error:
```
TSTypes.scala: Missing implicit for default TSType[Seq[ListObj]]. Bring it in scope with `import DefaultTSTypes._` or `extends DefaultTSTypes`
```
(The error doesn't change if i `import DefaultTSTypes._` or do `object TSTypes extends DefaultTSTypes`).

**Comment by dhoepelman** [comment]
Hi @jnatten

Your definitions should work. The immidiate problem is probably that you're not importing your custom definitions into the generator. Add `typescriptGenerationImports := Seq("your.package.TSTypes._")`. I will look into making this error a little clearer.

More interesting is that the behavior you are looking for is supposed to work out of the box.
[Trying to reproduce your problem](https://github.com/scala-tsi/scala-tsi/pull/179) I narrowed it down to you needing this:
```scala
object TSTypes {
  implicit val listObj: TSIType[ListObj]             = TSType.fromCaseClass[ListObj]
  implicit val coolEnumValue: TSType[CoolEnum.Value] = TSType.sameAs[CoolEnum.Value, CoolEnum.type]
}
```

There's 2 things here:
1. Bug in enum support

`CoolEnum.type` is automatically supported, but `CoolEnum.Value` isn't.
This oversight needs to be manually corrected, by telling scala-tsi to same thing for `MyEnum.Value` as for `MyEnum.type`

2. the default generation isn't used for `ListObj`

You shouldn't need to specify this, as fromCaseClass is the default for a case class like ListObj. I think this is caused by having a custom definition in the mix (because of bug 1), but even in that case it should still use the default.

I'll look into it, in the mean time you can use the supplied workaround.


Unrelated, if I may give some unsollicited advice, Scala 2 Enumerations are a bit meh. I can highly recommend [enumeratum](https://github.com/lloydmeta/enumeratum). Here's [a TSType for it](https://github.com/scala-tsi/scala-tsi/issues/74#issuecomment-648803395)

**Comment by jnatten** [comment]
I might not have been clear enough, but we already did import the custom definitions (and those work). I was just hoping to not have to do `TSType.fromCaseClass[A]` for every type that was used in a list.

I have experienced the same without the enum entirely.

I still have to do:
```scala
case class ListObj(a: String, b: String)

case class ParentObj(
  coolList: Seq[ListObj]
)
```
```scala
import com.scalatsi._
import com.scalatsi.dsl._

object TSTypes {
  implicit val listObj: TSType[ListObj] = TSType.fromCaseClass[ListObj]
}
``` 
Even if the enum isn't present.
It isn't a _big_ deal, but it sure gets to be a lot of types that need specifying after a while, and if its possible to detect them automagically it would be awesome.

Our project is open source, so if it could help to look at one of our repositories with scala-tsi, [here](https://github.com/NDLANO/draft-api) is one.
Only files of interest are probably:
[build.sbt](https://github.com/NDLANO/draft-api/blob/08490463f13882e31fa632b948310804f984fffc/build.sbt#L109)
[TSTypes.scala](https://github.com/NDLANO/draft-api/blob/08490463f13882e31fa632b948310804f984fffc/src/main/scala/no/ndla/draftapi/model/api/TSTypes.scala#L18)
[Article.scala](https://github.com/NDLANO/draft-api/blob/08490463f13882e31fa632b948310804f984fffc/src/main/scala/no/ndla/draftapi/model/api/Article.scala#L27) (One of the exported types. The line in the url is one of the types that require a `TSType.fromCaseClass[RequiredLibrary]` to work when in a `Seq`)

Your enum workaround was a lot cleaner than ours, I'll definitely use that instead, thanks!

**Comment by dhoepelman** [comment]
Yep you're right. Container types like `Seq[T]`, `Either[L,R]` or `Map[String, T]` don't work by default if `T` is generated by the macros, like for case class. This will require some tinkering with the macro.


#### Investigation notes

The problem is with the case class macro in `Macros.generateInterfaceFromCaseClass`. For `case Class(field: Seq[InList])` currently it generates the following code:

```
ListMap(
  ("field", TSType.getOrGenerate[Seq[InList]])
)
```

The `TSType.getOrGenerate[Seq[InList]]` comes from `Macros.getTSType`.

In the case of a parameterized type like `Seq`, we have a implicit def available for `Seq`, and can generate one for `InList`, but `Seq`'s implicit def can't be used because there is not implicit in scope (yet) for `InList`.

`Macros.getTSType` should be modified to instead bring the type parameters' `TSType` in implicit scope, for a parameterized types like `Paramaterized[T1, T2, ...]`, output:
```
{
    implicit val typeParam1: TSType[T1] = TSType.getOrGenerate[T1]
    implicit val typeParam2: TSType[T2] = TSType.getOrGenerate[T2]
    // ...
    TSType.getOrGenerate[Parameterized[T1, T2, ...]]
}
```

For `case Class(field: Seq[InList])`, this will result in:
```
ListMap(
  ("field", {
    implicit val typeParam1: TSType[InList] = TSType.getOrGenerate[InList]
    // Seq's implicit def can now be used, because TSType[InList] is in scope
    TSType.getOrGenerate[Seq[InList]]
  })
)
```

thus fixing the need to manually need to define a `implicit val ...: TSType[InList] = TSType.fromCaseClass[InList]` like from the workaround.

This will also improve the error message for any genuine missing implicits (if `TSType[InList]` is not defined and cannot be generated), as it will now complain about `InList` instead of `Seq[InList]`.

**Comment by dhoepelman** [comment]
@jnatten a 0.5.1-SNAPSHOT with a fix for the case classes required should be in maven central soon, give it a whirl.

I am a bit concerned about performance, could you report the time `sbt clean generateTypescript` takes in your project on 0.5.0, and on 0.5.1-SNAPSHOT?

The enumeration issue turns out to be a bit trickier (did I mention Scala 2 enums suck😉 ?). Basically there's no easy way to get from `CoolEnum.Value` to `CoolEnum`, so you'll have to keep the `sameAs` for now.

**Comment by jnatten** [comment]
I wasn't able to find the snapshot version on mavencentral, but i did a `sbt publishLocal` from the [0.5.1-m1](https://github.com/scala-tsi/scala-tsi/tree/0.5.1-m1) tag, and the performance didn't really seem impacted at all (We don't have that many exported types in each repository though).

It took 19 seconds with both versions (with and without `fromCaseClass` in the snapshot). So for our case the performance isn't really impacted at all.

I found that the `fromCaseClass` is still required for types used in `Either`'s, but this is still a big improvement for us!
Thank you for your work, it is awesome!

**Comment by dhoepelman** [comment]
Can you provide an example of where it's still neccessary for Either? It should work.

**Comment by jnatten** [comment]
I figured out the "problem". We had a type alias for `Either[SomeType, Long]` which was used in a `Seq` that didn't work.
When i directly used the type instead of the type alias it worked perfectly.
Not critical for us, but the alias is defined in [package object](https://github.com/NDLANO/draft-api/blob/master/src/main/scala/no/ndla/draftapi/model/api/package.scala#L12) and used in the [Article](https://github.com/NDLANO/draft-api/blob/master/src/main/scala/no/ndla/draftapi/model/api/Article.scala#L43) type here. The article type is the one that is exported.

**Comment by dhoepelman** [comment]
Need to look further into how type aliases work in macro's. Shelving that for now.

**Comment by dhoepelman** [comment]
Fix released in 0.5.1
